src/control/ToolHandler.cpp

```diff
@@ -105,8 +108,13 @@ ToolHandler::~ToolHandler() {
     this->settings = nullptr;
 }
 
-void ToolHandler::setEraserType(EraserType eraserType) {
-    this->eraserType = eraserType;
+void ToolHandler::initButtonTool(ToolPointer tp, ToolType type) {
+    setToolPointer(new Tool(tools[type - TOOL_PEN].get()), tp);
+}
```

Comment:
I think this should rather be something like:

```cpp
Tool* tool = getToolPointer(tp);
delete tool;
setToolPointer(new Tool(tools[type - TOOL_PEN].get()), tp);
```

Otherwise, this would be a memory leak. However, right now this doesn't work because the "copy constructor" `Tool(Tool* t)` doesn't do a deepcopy of the `thickness` array (which is a member of `Tool`). 

I would see two solutions:
1. Include deepcopy of `thickness` array in copy constructor: I tried this but for whatever reason it did crash the program in other parts (probably a programming mistake but somehow this solution also feels ugly).
2. Use `unique_ptr`'s for the the button pointers rather than normal pointers: This sounds like the right solution to me but I didn't use smart pointers before, so I'm a little bit unsure how to approach this.

@Febbe , @LittleHuba : What would be the correct way to approach this? If you could give me pointer on how to do it either way that would help me a lot.

Reply:
Little update:

I managed to implement the `unique_ptr` and the `deepcopy` :tada: